### ISSUE TITLE
Use async HttpClient in preference to HttpWebRequest for improved parallel performance

### ISSUE THREAD

**Issue by davesmits** [dotnetrdf/dotnetrdf#175]
hi,

in our application we are facing some some performance issues and we been able to track it down to dotnetrdf. I created a unit test to explain the problem. The unit test can be found here: https://github.com/davesmits/dotnetrdf/blob/master/Dave.Tests/UnitTest1.cs

The unit tests expects jena/fuseki is running localhost:3030 with a datastore Test.

What happens:

i try to run 30 sparql queries in parallel. When doing this via dotnetrdf its almost like each new call get a second delay. Running these 30 queries results in a average of +- 15 seconds and the longest running one is about 26/27 seconds.

To figure out if it was dotnetrdf or Jena I build also a test that uses the HttpClient to query Jena directly. In this test i am skipping the parsing of the sparqlresult, but the results are the calls take a average of just 0.5 seconds. I understand parsing the result takes some time too, 14 seconds difference is a HUGE difference.

Any advice what we can do to get better performance?

**Comment by kal** [comment]
As you point out, you are not really comparing apples and apples as you skip result parsing or indeed any other sort of access to the result content. Its not clear from the test how big these result sets are, so it could be just that you are pulling back a lot of data. 

It would be helpful if you could pin down more precisely where the performance hit lies by doing some profiling of your test code.

**Comment by davesmits** [comment]
Ohh and btw. I now have a implementation that is doing the parsing as well. THere is no big difference in performance with the http client implementation in the unit test. 

So you can really blame the HttpRequests
````
private async Task<object> QueryAsync(string sparqlQuery)
        {            
            SparqlResultSet results = new SparqlResultSet();

            var request = new HttpRequestMessage(HttpMethod.Post, new Uri(this._queryUri));
            request.Headers.TryAddWithoutValidation("Accept", MimeTypesHelper.HttpRdfOrSparqlAcceptHeader);
            request.Content = new FormUrlEncodedContent(new [] { new KeyValuePair<string, string>("query", sparqlQuery) });

            using (var client = new HttpClient())
            {
                using (var response = await client.SendAsync(request))
                {
                    response.EnsureSuccessStatusCode();
                    using (var responseStream = await response.Content.ReadAsStreamAsync())
                    {
                        using (var streamReader = new StreamReader(responseStream))
                        {
                            Graph g = new Graph();
                            string ctype = response.Content.Headers.ContentType.MediaType;
                            try
                            {
                                
                                IRdfHandler rdfHandler = new GraphHandler(g);
                                // Is the Content Type referring to a RDF format?
                                IRdfReader rdfreader = MimeTypesHelper.GetParser(ctype);
                                rdfreader.Load(rdfHandler, streamReader);
                            }
                            catch (RdfParserSelectionException)
                            {
                                ISparqlResultsHandler resultsetHandler = new ResultSetHandler(results);
                                // If we get a Parser selection exception then the Content Type isn't valid for a RDF Graph

                                // Is the Content Type referring to a Sparql Result Set format?
                                ISparqlResultsReader resreader = MimeTypesHelper.GetSparqlParser(ctype, true);
                                resreader.Load(resultsetHandler, streamReader);
                            }

                            if (results.ResultsType != SparqlResultsType.Unknown)
                            {
                                return results;
                            }
                            else
                            {
                                return g;
                            }
                        }
                    }
                }
            }
        }
````

**Comment by kal** [comment]
Thanks @davesmits , that's much more useful information ;-)

It really would be good to move DNR away from the "legacy" HttpWebRequest stuff and over to HttpClient, but that looks like quite a bit of work to support properly. However it seems like in the .NET Core world we really do need to get this done because it affects not only this ticket but also the timeout issues reported on #172. The problem will be maintaining compatibility with the really old versions of the .NET Framework that some people (especially those using Unity) are stuck with.

We already have #88 open for this in the 3.0 milestone, but I think the question here is practically what can we do in the 2.x timeframe given that 2.x is still supporting all those old framework versions. @tpluscode do you have any thoughts on this?

**Comment by tpluscode** [comment]
Hm. so we have the `HttpWebRequest` implementation which works with all versions but is not included in .NET Standard. On the other hand `HttpClient` is supported on all Standard but only on .NET 4.5+.

Do I reason correctly that we should ideally move to `HttpClient` on all targets except .NET 4.0 and start using async/await to get parallelism with web requests?

I see how this could be a big task but probably a worthy goal for a hypothetical 2.5 release?

**Comment by kal** [comment]
Thinking some more about this, my *preferred* solution would be to defer to 3.0. However I'm pretty sure that getting to 3.0 is going to take quite a lot more effort and without a long dedicated chunk of time (which I don't have right now) is not going to happen soon enough.

So I think it probably should be a goal for a 2.x release to move to HttpClient for everything except .NET40. We may as well keep this ticket for that task as it gives all of the history. I'm just going to rename the ticket to make it clearer

**Comment by davesmits** [comment]
we use dotnetrdf in our webapi project; but had even more issues. 

I figured out that parsing of a the result from jena via de dotnetrdf code is "heavy". deserialization easy takes up over a second and let the cpu spike. Running load tests with 25 concurrent users easily took the webapi down because of this. number of triples were about +- 1000 triples.

this is the code in dotnetrdf causing this delays:

````
try
                            {
                                
                                IRdfHandler rdfHandler = new GraphHandler(g);
                                // Is the Content Type referring to a RDF format?
                                IRdfReader rdfreader = MimeTypesHelper.GetParser(ctype);
                                rdfreader.Load(rdfHandler, streamReader);
                            }
                            catch (RdfParserSelectionException)
                            {
                                ISparqlResultsHandler resultsetHandler = new ResultSetHandler(results);
                                // If we get a Parser selection exception then the Content Type isn't valid for a RDF Graph

                                // Is the Content Type referring to a Sparql Result Set format?
                                ISparqlResultsReader resreader = MimeTypesHelper.GetSparqlParser(ctype, true);
                                resreader.Load(resultsetHandler, streamReader);
                            }
````
I replaced this code by using deserialization via json.net and manually convert everything to triples manully. This isnt the full support as this code above has but the performance penalty for us was too big. Just something to consider

**Comment by kal** [comment]
There are the places where HttpClient is used in the core library:

- `NetCoreCompatibility` - provides synchronous wrappers around the async methods in .NET Core's HttpWebRequest API. Should be possible to remove this after the rest of the classes are updated
- `BaseEndpoint.ApplyCustomRequestOptions()` : HttpWebRequest is exposed in the public API, so changing this would be a **breaking** API change for .NET Core.
- `Tools.HttpDebugRequest()`: Another place where the public API exposes HttpWebRequest, so another **breaking** API change
- `Parsing.UriLoader` : Internal use in the class
- `Parsing.UriLoaderAsync`: Internal use in the class
- `Query.Inference.Pellet.Services.*`: The service classes all use HttpWebRequest internally
- `Query.Inference.Pellet.PelletServer`: Also uses HttpWebRequest internally
- `Query.SPARQLRemoteEndpoint`: Internal use
- `Storage.Management.AllegroGraphServer`: Internal
- `Storage.Management.SesameServer`: internal
- `Storage.Management.StardogServer`: mostly internal, BaseStardogServer does expose HttpWebRequest in a protected virtual method so this would be a **breaking** api change, though quite a minor one
- `Storage.Management.BaseHttpConnector`: Again protected methods in a public base class, so a **breaking** api change
- `Storage.Management.SesameHttpProtocolConnector`: Protected methods in a public base class, so a **breaking** api change
- 'Storage.Management.*Connector`: The above base class methods are overridden in most of the connector classes - so more breaking API changes there
- `Storage.SparqlHttpProtocolConnector`: Used internally
- `Update.SparqlRemoteUpdateEndpoint`: Used internally

@tpluscode - my concern here is that because there are some public APIs that expose `HttpWebRequest` we would have to introduce breaking API changes to support `HttpClient` properly. 

So given this and the effort that would be involved, I'm going back to my original position that this should be deferred to the next major release.
